run-tests.php

```diff
@@ -1224,6 +1268,261 @@ function run_all_tests($test_files, $env, $redir_tested = null)
+                                proc_terminate($workerProcs[$i]);
+                                unset($workerProcs[$i]);
+                                unset($workerStdins[$i]);
+                                unset($workerStdouts[$i]);
+                                unset($workerStderrs[$i]);
+                                goto escape;
+                            }
+                            break;
+                        case "test_result":
+                            [$name, $index, $result, $resultText] = [$message["name"], $message["index"], $message["result"], $message["text"]];
+                            foreach ($message["PHP_FAILED_TESTS"] as $category => $tests) {
+                                $PHP_FAILED_TESTS[$category] = array_merge($PHP_FAILED_TESTS[$category], $tests);
+                            }
+                            $test_idx++;
+                            $testsInProgress--;
+                            clear_show_test();
+                            echo $resultText;
+                            show_test($test_idx, "⚡️[" . count($workerProcs) . "/$workers concurrent test workers running]⚡️");
+
+                            if (!is_array($name) && $result != 'REDIR') {
+                                $test_results[$index] = $result;
+
+                                if ($failed_tests_file && ($result == 'XFAILED' || $result == 'FAILED' || $result == 'WARNED' || $result == 'LEAKED')) {
+                                    fwrite($failed_tests_file, "$index\n");
+                                }
+                                if ($result_tests_file) {
+                                    fwrite($result_tests_file, "$result\t$index\n");
+                                }
+                            }
+                            break;
+                        case "error":
+                            kill_children($workerProcs);
+                            error("Worker $i reported error: $message[msg]");
+                            break;
+                        default:
+                            kill_children($workerProcs);
+                            error("Unrecognised message type '$message[type]' from worker $i");
+                    }
+                }
+            }
+        }
+escape:
+    }
+
+    clear_show_test();
+
+    if ($test_idx !== $test_cnt) {
+        error("Somehow, " . ($test_cnt - $test_idx) . " tests never got executed. This is probably a bug in parallel test execution.");
+    }
+    kill_children($workerProcs);
+}
+
+function send_message($stream, array $message) {
+    fwrite($stream, base64_encode(serialize($message)) . "\n");
+}
+
+function kill_children(array $children) {
+    foreach ($children as $child) {
+        if ($child) {
+            proc_terminate($child);
+        }
+    }
+}
+
+function run_worker() {
+    global $workerID;
+
+    @unlink(__DIR__ . "/../worker$workerID.log");
+    ini_set("error_log", __DIR__ . "/../worker$workerID.log");
```

Comment:
I forgot to remove debug code here. Probably because it never worked in the first place, maybe I need to touchback an additional INI setting.

I'd like to add an error handler that sends a message with the error.